pkg/domain/infosync/info.go

```diff
@@ -501,13 +511,67 @@ func doRequest(ctx context.Context, apiName string, addrs []string, route, metho
 				return nil, err
 			}
 			if res.StatusCode != http.StatusOK {
-				logutil.BgLogger().Warn("response not 200",
+				var (
+					leaderAddr string
+					errMsgMap  = map[string]string{}
+				)
+				for _, addr2 := range addrs {
+					status, err2 := etcdCli.Status(ctx, addr2)
+					if err2 != nil {
+						errMsgMap[addr2] = err2.Error()
+						continue
+					}
+					if status.Leader == status.Header.MemberId {
+						leaderAddr = addr2
+						break
+					}
+				}
+
+				errMsgField := zap.Skip()
+				if len(errMsgMap) > 0 {
+					errMsgField = zap.Any("errors when find leader", errMsgMap)
+				}
+
+				logutil.BgLogger().Warn("response not 200, will retry sending to etcd leader",
 					zap.String("method", method),
 					zap.String("hosts", addr),
 					zap.String("url", url),
 					zap.Int("http status", res.StatusCode),
 					zap.Int("address order", idx),
+					zap.String("leader", leaderAddr),
+					errMsgField,
 				)
+				leaderIsUsed := leaderAddr == addr
+				if intest.InTest {
+					leaderIsUsed = false
+				}
+
+				if leaderAddr != "" && !leaderIsUsed {
+					url2 := util2.ComposeURL(leaderAddr, route)
+					req2, err2 := http.NewRequestWithContext(ctx, method, url2, bytes.NewReader(body))
+					if err2 != nil {
+						return nil, err2
+					}
+					if body != nil {
+						req2.Header.Set("Content-Type", "application/json")
+					}
+					res2, err2 := doRequestWithFailpoint(req2)
+					if err2 == nil {
+						bodyBytes2, err2 := io.ReadAll(res2.Body)
+						terror.Log(res2.Body.Close())
+						if err2 != nil {
+							return nil, err2
```

Comment:
should we keep try other addresses here like previous err2?

Reply:
By the design we only try to connect PD leader (the same as etcd leader) to solve the oncall issue.